### PR TITLE
#255: Implemented delete route and the tests. Added the puzzle to com…

### PR DESCRIPTION
**Pull request by ekondrashev**
#255 Implemented delete route and the tests. 
* Added the puzzle to complete with edit route and the remainder from the original puzzle.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#313](https://github.com/timelesslounge/timelessis/issues/313) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#313](https://github.com/timelesslounge/timelessis/issues/313) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/CF9LG1YCA/7951eab5-5741-42bb-a187-894d4e245786); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: tests/integration/it_roles_test.py [hunk 1]
```diff
@@ -9,20 +9,44 @@ def test_list(client):
...
 
 
 def test_edit(client):
-    assert client.get("/roles/edit/1").status_code == HTTPStatus.OK
+    assert client.get(url_for('role.edit', id=1)).status_code == HTTPStatus.NOT_FOUND
+
+
+def test_delete_not_found(client):
+    assert Role.query.filter_by(id=1).count() == 0
+    result = client.post(url_for('role.delete', id=1))
```
Comment (amihaiemil) [inline tests/integration/it_roles_test.py]: @ekondrashev can we inline this variable? It's only used once

File: timeless/roles/views.py [hunk 2]
```diff
@@ -33,6 +36,9 @@ def create():
 
 @bp.route("/edit/<int:id>", methods=("GET", "POST"))
 def edit(id):
+    table = Role.query.get(id)
+    if not table:
+        return abort(HTTPStatus.NOT_FOUND)
```
Comment (amihaiemil) [inline timeless/roles/views.py]: @ekondrashev this ``if`` should go in the next one, right? We should only try to read the table if it's a ``POST``.

File: timeless/roles/views.py [hunk 3]
```diff
@@ -46,7 +52,12 @@ def edit(id):
...
-def delete():
-    flash("Delete not yet implemented")
+@bp.route("/delete/<int:id>", methods=["POST"])
+def delete(id):
+    roles = Role.query.get(id)
+    if not roles:
+        return abort(HTTPStatus.NOT_FOUND)
+    roles = Role.query.get(id)
+    DB.session.delete(roles)
+    DB.session.commit()
```
Comment (amihaiemil) [inline timeless/roles/views.py]: @ekondrashev I don't like having the ``DB`` in ``views.py``. See how ``CrudAPIView`` is used in other places, that one was implemented to decouple DB from the views, as far as I understood :)

File: tests/integration/it_roles_test.py [hunk 4]
```diff
@@ -9,20 +9,44 @@ def test_list(client):
...
-    })
-    assert response.location.endswith(url_for('role.list_roles'))
-    assert Role.query.count() == 1
-    assert Role.query.get(1).name == "owner"
+def test_create(client, db_session):
+    assert Role.query.count() == 0
+    created = None
+    try:
+        response = client.post(url_for("role.create"), data={
+            "name": "owner"
```
Comment (amihaiemil) [inline tests/integration/it_roles_test.py]: @ekondrashev ``name``, ``owner`` of what exactly? I would say ``normal_user`` and ``john``-- I suppose the ``owner`` is the employee who has said role, or?

Comment (ekondrashev) [reply tests/integration/it_roles_test.py]: @amihaiemil I'm not an author not sure what was original thought here, I'll change this to `John`

**Comment by ekondrashev** [comment]
@amihaiemil done addressing the comments, please take another look

File: tests/integration/it_roles_test.py (hunk 1 above)
Comment (ekondrashev) [reply tests/integration/it_roles_test.py]: @amihaiemil done, thx

File: timeless/roles/views.py (hunk 2 above)
Comment (ekondrashev) [reply timeless/roles/views.py]: @amihaiemil right, fixed, thx!

File: timeless/roles/views.py (hunk 3 above)
Comment (ekondrashev) [reply timeless/roles/views.py]: @amihaiemil added a puzzle to address this

**Comment by amihaiemil** [comment]
@ekondrashev looks good, thanks!

**Comment by amihaiemil** [comment]
@rultor good to merge

File: timeless/roles/views.py (hunk 3 above)
Comment (emilianodellacasa) [reply timeless/roles/views.py]: @ekondrashev `CrudAPIView`  shouldn't be used for `Role`, as it is not meant to be returned with a REST api, Better refer to timeless/companies/views.py file to see how it should be implemented. Please change both puzzles accordingly

**Comment by emilianodellacasa** [comment]
@ekondrashev please see my comment

**Comment by ekondrashev** [comment]
@emilianodellacasa done, pls lmk if it is better now

**Comment by emilianodellacasa** [comment]
@ekondrashev Thanks, now travis is failing because of pylint, check the log please

**Comment by ekondrashev** [comment]
@emilianodellacasa seems like some issues with `pylint` trying to figure what happened

**Comment by emilianodellacasa** [comment]
@ekondrashev Still something to do for pylint

**Comment by ekondrashev** [comment]
@emilianodellacasa I've removed `DB.session` usage in favor of calling delete method for found Role. Pylint was complaining because of not being able to find `commit` and `delete` methods for `scoped_session` instance. I've tried to add the ignorance for `scoped_session` class with `.pylintrc` but seems like `check_pylint.py` script does not take it into account

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job was finished in 22 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#313](https://github.com/timelesslounge/timelessis/issues/313) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa)

**Comment by ypshenychka** [comment]
@ekondrashev According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>Pull request title explains the problem it is fixing

Please provide brief and clear title. Right now it's an unfinished sentence.

**Comment by ekondrashev** [comment]
@ypshenychka done, thank you

**Comment by ypshenychka** [comment]
@ekondrashev thanks

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by amihaiemil** [comment]
@ypshenychka pls repeat the verdict, 0crat seems to have missed it :(

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +20 point(s) just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Comment by ypshenychka** [comment]
@amihaiemil I've received payment. Didn't you?

**Comment by amihaiemil** [comment]
@ypshenychka yes, I did, right after I pinged you. Thanks! :)

**Pull request merged**